libcontainer/seccomp/seccomp_linux.go

```diff
@@ -60,14 +72,14 @@ func InitSeccomp(config *configs.Seccomp) (int, error) {
 			// filter). We will be blocked on read()/close() inside syncParentSeccomp() but if the seccomp
 			// agent allows those syscalls to proceed, initialization works just fine and the agent can
 			// handle future read()/close() syscalls as it wanted.
-			if call.Name == "write" {
+			if call.Name == writeSyscall {
 				return -1, errors.New("SCMP_ACT_NOTIFY cannot be used for the write syscall")
 			}
 		}
 	}
 
-	// See comment on why write is not allowed. The same reason applies, as this can mean handling write too.
-	if defaultAction == libseccomp.ActNotify {
+	// See comment on why write is not allowed. The same reason may apply if writeAllowed is false, as this can mean handling write too.
+	if !writeAllowed && defaultAction == libseccomp.ActNotify {
```

Comment:
I think this is buggy, the default action can still be a write and we should deny it. Just because we don't know what path of all the seccomp rules will be used when runc issues a write.

But leaving that technicality aside, there seems to be a deeper issue that is not easy to fix. The issue is: seccomp profiles allows to do complex rules to match when a syscall will be allowed and when they won't. I know most profiles are just a list of allow/deny syscalls (although moby profile is not trivial), but the profile allows _much_ more than that and there is where this becomes complex. With that flexibility we can't easily answer if "write will always be allowed" easily and reliably.

Let me elaborate.

The logic checking if write is allowed here is quite simple, and is not good enough to know if a write call will be _always_ be allowed. The seccomp profile let's you express lot of subtleties (like if the param to the syscall is this or the other), it can totally be that the write is allowed when the fd is 0 and that will make this container not start and hang in a very hard to debug way. This logic doesn't detect that.

Detecting this reliably is tricky, but I _think_ we can make heuristics. We can check if args is not set, and in that case if the write is unconditionally allowed it might work.

I would like to think and have an eye from others about this fact, though.

What do you think?

Reply:
Another option, rather than heuristics, would be to try to use libseccomp to know if a syscall will always be allowed. I don't know if they have such a thing in their API, but if they do we could solve the problem :)

Reply:
@kolyshkin is a maintainer, might now... ;)